cocotb/drivers/amba.py

```diff
@@ -259,63 +259,71 @@ def __init__(self, entity, name, clock, memory, callback=None, event=None,
 
         cocotb.fork(self._read_data())
-        cocotb.fork(self._write_data())
+        cocotb.fork(self._write_transaction())
 
     def _size_to_bytes_in_beat(self, AxSIZE):
         if AxSIZE < 7:
             return 2 ** AxSIZE
         return None
 
     @cocotb.coroutine
-    def _write_data(self):
-        clock_re = RisingEdge(self.clock)
+    def _write_address(self):
+        """Monitor the Write Address channel for transactions
+        """
+        self.bus.AWREADY <= 1
+        #TODO maybe some configurable delays
+        while not self.bus.AWVALID.value:
+            yield RisingEdge(self.clock)
+        self.bus.AWREADY <= 0
 
-        while True:
-            while True:
-                self.bus.WREADY <= 0
-                yield ReadOnly()
-                if self.bus.AWVALID.value:
-                    self.bus.WREADY <= 1
-                    break
-                yield clock_re
+        awaddr = int(self.bus.AWADDR)
+        awlen = int(self.bus.AWLEN)
+        awsize = int(self.bus.AWSIZE)
+        awburst = int(self.bus.AWBURST)
+        awprot = int(self.bus.AWPROT)
+        return awaddr, awlen, awsize, awburst, awprot
 
-            yield ReadOnly()
-            _awaddr = int(self.bus.AWADDR)
-            _awlen = int(self.bus.AWLEN)
-            _awsize = int(self.bus.AWSIZE)
-            _awburst = int(self.bus.AWBURST)
-            _awprot = int(self.bus.AWPROT)
+    @cocotb.coroutine
+    def _write_data(self):
+        """Monitor the Write Address channel for transactions
+        """
+        self.bus.WREADY <= 1
+        while not self.bus.WVALID.value:
+            yield RisingEdge(self.clock)
+        #TODO support bursts
+        self.bus.WREADY <= 0
 
-            burst_length = _awlen + 1
-            bytes_in_beat = self._size_to_bytes_in_beat(_awsize)
+        wdata = int(self.bus.WDATA)
+        wlast = int(self.bus.WLAST)
+        wid = int(self.bus.WID)
+        wstrb = int(self.bus.WSTRB)
+        return wdata, wid, wstrb
 
-            word = BinaryValue(n_bits=bytes_in_beat*8, bigEndian=self.big_endian)
+    @cocotb.coroutine
+    def _write_transaction(self):
+        """Process write transactions
 
-            if __debug__:
-                self.log.debug(
-                    "AWADDR  %d\n" % _awaddr +
-                    "AWLEN   %d\n" % _awlen +
-                    "AWSIZE  %d\n" % _awsize +
-                    "AWBURST %d\n" % _awburst +
-                    "BURST_LENGTH %d\n" % burst_length +
-                    "Bytes in beat %d\n" % bytes_in_beat)
+        This requires both wdata and awaddr to complete.
+        TODO support out of order threads
+        """
+        while True:
+            self.bus.BVALID <= 0
+            d = cocotb.fork(self._write_data())
+            a = cocotb.fork(self._write_address())
+            wdata, wid, wstrb = yield d.join()
+            awaddr, awlen, awsize, awburst, awprot = yield a.join()
```

Comment:
I think you can use `Combine` here, as

```
(
    (wdata, wid, wstrb),
    (awaddr, awlen, awsize, awburst, awprot)
) = yield Combine(
    self._write_data(),
    self._write_address()
)

Reply:
Yes!... which is of course on master :P I was looking for such a thing all afternoon.
I guess I'll figure out my other issues and rebase